File: Decoder/decoder.php
<?php

if ($_SERVER["REQUEST_METHOD"] == "POST") {
  // Reading existing passwords and encrypting
  $existingPasswords = readPasswords('passwords.txt');
  // Reading user's encrypted passwords
  $userPasswords = readUsers('users.csv');
  // Creating hashes before comparing.
  $hashedExistingPasswords = creatingHashes($existingPasswords);
  // Comparing passwords
  comparingPasswords($userPasswords, $hashedExistingPasswords);
}

function creatingHashes($existingPasswords) {
  $hashesArray = array();
  // 1 word password
  foreach ($existingPasswords as $existingPassword1) {
    $hashedPassword = hash('sha256', $existingPassword1);
    // Adding 1 word password to array
    array_push($hashesArray, trim($hashedPassword));
    // 2 word password
    foreach ($existingPasswords as $existingPassword2) {
      $combinedPassword1 = $existingPassword1 . $existingPassword2;
      $hashedPassword2 = hash('sha256', $combinedPassword1);
      // Adding 2 words password to array
      array_push($hashesArray, trim($hashedPassword2));
      // 3 word password
      foreach ($existingPasswords as $existingPassword3) {
        $combinedPassword2 = $existingPassword1 . $existingPassword2 . $existingPassword3;
        $hashedPassword3 = hash('sha256', $combinedPassword2);
        // Adding 2 words password to array
        array_push($hashesArray, trim($hashedPassword3));
      }
    }
  }
  return $hashesArray;
}

function readPasswords($fileName) {
  $passwords = array();
  
  // Comprobar si el archivo existe
  if (!file_exists($fileName)) {
      throw new Exception("El archivo no existe");
  }

  // Comprobar si el archivo se puede abrir
  if (!is_readable($fileName)) {
      throw new Exception("El archivo no se puede leer");
  }

  // Abre el archivo TXT
  $file = fopen($fileName, "r");
  
  // Itera por todas las líneas del archivo
  while (($linea = fgets($file)) !== false) {
      // Agrega al arreglo la palabras
      array_push($passwords, trim($linea));
  }

  // Cierra el archivo TXT
  fclose($file);

  return $passwords;
}

function readUsers($fileName) {
  // Open the CSV file for reading
  $fileOpened = fopen($fileName, 'r');

  // Check if the file was successfully opened
  if ($fileOpened !== false) {
      // Initialize an empty array to store the passwords
      $passwords = array();

      // Read each line of the CSV file
      while (($data = fgetcsv($fileOpened)) !== false) {
          if(count($data) >= 2){
              $username = $data[0];
              array_push($passwords, trim($data[1]));
          }
      }
      // Close the CSV file
      fclose($fileOpened);
    } else {
      echo 'Archivo no se ha podido abrir';
    }
    return $passwords;
}

function comparingPasswords($userPasswords, $hashedExistingPasswords) {
  echo '<h4>Comparando hashes:<h4>';
  // Iterating through user passwords
  foreach ($userPasswords as $userPassword) {
    foreach ($hashedExistingPasswords as $hashedPassword) {
      // Check the flag after the inner loop
      if ($hashedPassword == $userPassword) {
        echo 'Exito: ' . $hashedPassword . ' == ' . $userPassword . '<br>';
      }
    }
  }
}

?>

File: Decoder/decoder.py
import csv
from hashlib import sha256
import threading
from concurrent.futures import ThreadPoolExecutor
def read_passwords(file_name):
  passwords = []
  try:
    with open(file_name, 'r') as file:
      for line in file:
        passwords.append(line.strip())
  except FileNotFoundError:
    print("The file does not exist")
  except IOError:
    print("The file cannot be read")
  return passwords

def read_users(file_name):
  passwords = {}
  try:
    with open(file_name, 'r') as file:
      csv_reader = csv.reader(file, delimiter=';')
      for row in csv_reader:
        if len(row) >= 2:
          passwords[row[0]]= row[1].strip()
  except IOError:
    print("File could not be opened")
  return passwords

def creating_hashes(existing_passwords,start, end, hashes_map):
  for i in range(start, end):
    password1 = existing_passwords[i]
    print(i)
    hashed_password = sha256(password1.encode('utf-8')).hexdigest()
    hashes_map[password1] = hashed_password.strip()
    for j in range(len(existing_passwords)):
      password2 = existing_passwords[j]
      combined_password1 = password1 + password2
      hashed_password2 = sha256(combined_password1.encode('utf-8')).hexdigest()
      hashes_map[password2] = hashed_password2.strip()
      for k in range(len(existing_passwords)):
          password3 = existing_passwords[k]
          combined_password2 = password1 + password2 + password3
          hashed_password3 = sha256(combined_password2.encode('utf-8')).hexdigest()
          hashes_map[password3] = hashed_password3.strip()
  return hashes_map

def comparing_passwords(user_passwords, hashed_passwords):
  print('Comparing hashes:')
  for username in user_passwords.keys():
    for password in hashed_passwords.keys():
      if hashed_passwords[password] == user_passwords[username]:
        print(f'Success for {username} with password {password}:')
        print(f'{hashed_passwords[password]} == {user_passwords[username]}\n')

def main():
  num_threads = 12
  existing_passwords = read_passwords('passwords.txt')
  batch_size = len(existing_passwords) // num_threads
  user_passwords = read_users('users.csv')
  print(len(existing_passwords))
  result_dict = {}
  with ThreadPoolExecutor(max_workers=num_threads) as executor:
    futures = []

    for i in range(num_threads):
      start = i * batch_size
      end = start + batch_size if i < num_threads - 1 else len(existing_passwords)
      future = executor.submit(creating_hashes, existing_passwords, start, end, result_dict)
      futures.append(future)

    # Espera a que todos los hilos terminen
    for future in futures:
      future.result()
  comparing_passwords(user_passwords, result_dict)
  print("Fin") 

if __name__ == '__main__':
  main()
